llvm/lib/Target/AMDGPU/AMDGPUSwLowerLDS.cpp

```diff
@@ -0,0 +1,1333 @@
+    Value *BasePlusOffset =
+        IRB.CreateInBoundsGEP(IRB.getInt8Ty(), BaseLoad, {Offset});
+    LLVM_DEBUG(dbgs() << "Sw LDS Lowering, Replace non-kernel LDS for "
+                      << GV->getName());
+    replacesUsesOfGlobalInFunction(Func, GV, BasePlusOffset);
+  }
+  translateLDSMemoryOperationsToGlobalMemory(Func, LoadMallocPtr,
+                                             LDSInstructions);
+}
+
+static void reorderStaticDynamicIndirectLDSSet(KernelLDSParameters &LDSParams) {
+  // Sort Static, dynamic LDS globals which are either
+  // direct or indirect access on basis of name.
+  auto &DirectAccess = LDSParams.DirectAccess;
+  auto &IndirectAccess = LDSParams.IndirectAccess;
+  LDSParams.DirectAccess.StaticLDSGlobals = sortByName(
+      std::vector<GlobalVariable *>(DirectAccess.StaticLDSGlobals.begin(),
+                                    DirectAccess.StaticLDSGlobals.end()));
+  LDSParams.DirectAccess.DynamicLDSGlobals = sortByName(
+      std::vector<GlobalVariable *>(DirectAccess.DynamicLDSGlobals.begin(),
+                                    DirectAccess.DynamicLDSGlobals.end()));
+  LDSParams.IndirectAccess.StaticLDSGlobals = sortByName(
+      std::vector<GlobalVariable *>(IndirectAccess.StaticLDSGlobals.begin(),
+                                    IndirectAccess.StaticLDSGlobals.end()));
+  LDSParams.IndirectAccess.DynamicLDSGlobals = sortByName(
+      std::vector<GlobalVariable *>(IndirectAccess.DynamicLDSGlobals.begin(),
+                                    IndirectAccess.DynamicLDSGlobals.end()));
+}
+
+void AMDGPUSwLowerLDS::initAsanInfo() {
+  // Get Shadow mapping scale and offset.
+  unsigned LongSize =
+      M.getDataLayout().getPointerSizeInBits(AMDGPUAS::GLOBAL_ADDRESS);
+  uint64_t Offset;
+  int Scale;
+  bool OrShadowOffset;
+  llvm::getAddressSanitizerParams(Triple(AMDGPUTM.getTargetTriple()), LongSize,
+                                  false, &Offset, &Scale, &OrShadowOffset);
+  AsanInfo.Scale = Scale;
+  AsanInfo.Offset = Offset;
+  return;
+}
+
+bool AMDGPUSwLowerLDS::run() {
+  bool Changed = false;
+
+  // AddressSanitizer pass adds "nosanitize_address" module flag if it has
+  // instrumented the IR. Return early if the flag is not present.
+  if (!M.getModuleFlag("nosanitize_address"))
+    return Changed;
+
+  CallGraph CG = CallGraph(M);
+
+  Changed |= eliminateConstantExprUsesOfLDSFromAllInstructions(M);
+
+  // Get all the direct and indirect access of LDS for all the kernels.
+  LDSUsesInfoTy LDSUsesInfo = getTransitiveUsesOfLDS(CG, M);
+
+  // Utility to group LDS access into direct, indirect, static and dynamic.
+  auto PopulateKernelStaticDynamicLDS = [&](FunctionVariableMap &LDSAccesses,
+                                            bool DirectAccess) {
+    for (auto &K : LDSAccesses) {
+      Function *F = K.first;
+      if (!F || K.second.empty())
+        continue;
+
+      assert(isKernelLDS(F));
+      if (!F->hasFnAttribute(Attribute::SanitizeAddress))
```

Comment:
@JonChesterfield here if sanitize_address fn attribute is not present, this skips lowering of LDS.